src/hotspot/share/gc/g1/g1FullGCCompactionPoint.cpp

```diff
@@ -116,6 +116,38 @@ void G1FullGCCompactionPoint::add(HeapRegion* hr) {
   _compaction_regions->append(hr);
 }
 
-HeapRegion* G1FullGCCompactionPoint::remove_last() {
-  return _compaction_regions->pop();
+int G1FullGCCompactionPoint::compare(HeapRegion* const& hr, HeapRegion* const& e) {
+  return compare(hr->hrm_index(), e->hrm_index());
+}
+
+void G1FullGCCompactionPoint::sort_regions(){
+  regions()->sort(compare_ptr);
+}
+
+int G1FullGCCompactionPoint::find_sorted(HeapRegion* hr) {
+  bool found = false;
+  int pos = _compaction_regions->find_sorted<HeapRegion*, compare>(hr, found);
+  if (found) {
+    return pos;
+  }
+  return -1;
+}
+
+void G1FullGCCompactionPoint::move_regions_with_higher_hrm_index(G1FullGCCompactionPoint* cp, uint bottom) {
+  HeapRegion* cur = current_region();
+  assert(cur->hrm_index() >= bottom, "Sanity!");
+
+  HeapRegion* start = nullptr;
+  for (HeapRegion* r : *_compaction_regions) {
+    if (r->hrm_index() >= bottom) {
+      cp->add(r);
+      if (start == nullptr) {
+        start = r;
+      }
+    }
+  }
+
+  int start_index = find_sorted(start);
```

Comment:
Using regular iteration with an index, and potentially splitting it into two parts (one until `r->hrm_index() >= bottom`, and one afterwards) I think avoids the need for `find_sorted`.